action/protocol/staking/candidate_center.go

```diff
@@ -11,139 +11,347 @@ import (
 )
 
 type (
+	// CandidateView provides access to all candidates
+	CandidateView interface {
+		Size() int
+		All() (CandidateList, error)
+		Delta() CandidateList
+		Commit(CandidateList) error
+		ContainsName(string) bool
+		ContainsOwner(address.Address) bool
+		ContainsOperator(address.Address) bool
+		ContainsSelfStakingBucket(uint64) bool
+		GetByName(string) *Candidate
+		GetByOwner(address.Address) *Candidate
+		GetBySelfStakingIndex(uint64) *Candidate
+		Upsert(*Candidate) error
+	}
+
+	// candChange captures the change to candidates
+	candChange struct {
+		dirty map[string]*Candidate
+	}
+
 	// CandidateCenter is a struct to manage the candidates
 	CandidateCenter struct {
 		nameMap          map[string]*Candidate
 		ownerMap         map[string]*Candidate
 		operatorMap      map[string]*Candidate
 		selfStkBucketMap map[uint64]*Candidate
+		size             int
+		change           *candChange
```

Comment:
I didn't go into every detail. This may be an issue, because the changes are kept in candidate center as a struct level variable, instead of in dock.

Reply:
Dock is not meant to keep the dynamic changes at any time, but the final pending changes at the end of running Handler, which call ActionPostHandler(), which will capture all the changes and load into Dock, basically:

delta = all the changes to CandidateCenter
dock.Load( serialization of delta)

upon success of factory.Commit(), it will unload the confirmed change, and commit back

b = dock.Unload()
delta2 = b.Deserialize(), so it is identical to "delta" earlier
and then:
CandidateCenter.Commit(delta2)

tests are added in candidate_test.go to verify that CandidateCenter at time of generating "delta", and CandidateCenter after committing "delta2" are identical

Reply:
I have different understanding:
changes should be in dock, because dock goes with sm. In other word, each sm has its own changes, not the "inMemCC".